Initial draft defining syntax, semantics of controlling expressions (#65)
j3-fortran/fpp-202y: drafts/25-xxx-specifications.txt

5 Expressions allowed in #if and #elif directives
=================================================

ex05. Controlling expressions are made up of "primary expressions" and
      operators applied to subexpressions.

ex10. When evaluating a #if or #elif directive, preprocessing
      evaluates and expands all object-like macros and function-like
      macro calls to create a token-list of the expression to be
      evaluated.

ex15. The resulting list of tokens shall be a valid expression
      comprised of primary expressions and operators as described
      below.

ex15. Preprocessing computes the integer value of this expressions

>       using the greatest integer range available to the processor.to

Author: Of course. Thank you.

      determine the truth or falsity of the controlling expression.

> ex20. The evaluation of the expression shall not generate a
>       computational error (such as divide by zero).

Reviewer: Maybe we should be more explicit about what happens if a computational error occurs instead?
```suggestion
ex20. The processor shall reject a program if evaluation of
      the expression generates a computational error (such 
      as divide by zero).
```

ex20. When the expression evaluates to zero, the controlling expression
      will be considered "false". If the expression evaluates to
      any non-zero value, the controlling expression will be considered
      "true".

5.1 Primary expressions
-----------------------

Preprocessing recognizes the following primary expressions in
controlling expressions.

Since expression evaluation occurs *after* token expansion, there will
be no object-like macros or function-like macros left to evaluate. All
instances of ID or ID (args) will all have been replaced with their
expansions.

We list them here for completeness in describing the illustrative
syntax of controlling expressions.

| Primary      | Evaluation semantics                        |
|--------------+---------------------------------------------|
| ID           | The expansion of the object-like macro ID   |

> | ID (args)    | The expansion of the function-like macro ID |

Author: Let's only describe the post-expansion grammar, and not the pre-expansion grammar. There is no such pre-expansion grammar. Dan points out that undefined ID replacement has to be done *after* processing of `##` tokens.

| WHOLE_NUMBER | Decimal value of WHOLE_NUMBER               |
| ( expr )     | Parenthesized expressions                   |

> |--------------+---------------------------------------------|

Reviewer 2: Minor aside: C23 6.10.2-13 perversely also allows single-character character constants as "primaries" in conditional expressions, example from C23:
```
#if ’z’ - ’a’ == 25
```
however their exact CPP evaluation semantics are implementation-defined, which means their use is not guaranteed to be portable. I don't believe I've ever seen this bizarre "feature" used in practice. 

I suspect this is some weird legacy holdover in CPP and unless someone provides a strong rationale for their inclusion I think FPP should prohibit character constants in conditional expressions.

Reviewer: Agreed, character constants in conditional expressions should be a (pathological) processor-dependent extension.

controlling expressions in #if and #elif expressions are a subset of
those defined in the C 2023 standard §6.5 "Expressions" and §6.6
"Constant expressions".

A "precedence level" is assigned to each operator that determines
how the operators combine with sub-expressions containing other operators
at different precedence levels.

An "associativity" is assigned to each operator that determines how
operators at the same precedence level are combined.
    - "left" means the operator binds to the left,
    - "right" means the operator binds to the right.
    - "nonassoc" means that the operator is not associative.

The following table describes the semantics of the allowed operators
in conditional expressions. The table is grouped by precedence level,
from lowest precedence to highest.

We label subexpressions "e1", "e2", and "e3" to aid in describing the
evaluation semantics. Unless otherwise specified, all operators
evaluate with the same semantics as their Fortran counterparts.

| Prec |   Op    |    Syntax    |  Assoc'y |  Evaluation Semantics      |
|------+---------+--------------+----------+----------------------------|

> | low  |   ? :   | e1 ? e2 : e3 |  right   |      conditional-expr      |

Reviewer 2: I'll note in passing that CPP also allows the comma operator in conditional expressions (at lower priority than _conditional-expression_), although it's pretty pointless in preprocessor expressions and I'm not aware of any compelling use cases. 

For this reason we (implicitly) omitted it from the requirements doc in 25-114r2. I'm only raising it now in case someone has a compelling argument to include it (something other than strict compatibility with CPP), otherwise I'm fine dropping it.

Author: I left them out on purpose, but didn't have a strong reason to do so.

I may be missing a subtlety here.

In general, conditional expression evaluation is side-effect free. So, elaborating
```C
#if (my_complicated_expression, my_other_expression)
``` 
results in only the value of `my_other_expression` affecting the `#if`. `my_complicated_expression` may be evaluated, but its result is throw away.

Author: There's also the slippery slope in the C grammar, where the CPP conditional expressions start at _conditional-expression_, which I think can unfold down to _primary-expression_, which then includes the `(` _expression_ `)` which brings in the whole barnyard of comma-expressions and _assignment-expression_.

So I just chopped those rules out of the grammar.

Reviewer 2: We resolved to exclude the comma operator, which is now documented in dfc80

|------+---------+--------------+----------+----------------------------|
|      |   ¦¦    |   e1 || e2   |   left   |       Fortran .OR.         |
|------+---------+--------------+----------+----------------------------|
|      |   &&    |   e1 && e2   |   left   |       Fortran .AND.        |
|------+---------+--------------+----------+----------------------------|

> |      |    ¦    |   e1 | e2    |   left   |    Fortran IOR(e1, e2)     |
> |      |    ^    |   e1 ^ e2    |   left   |    Fortran IAND(e1, e2)    |
> |      |    &    |   e1 & e2    |   left   |    Fortran IEOR(e1, e2)    |

Reviewer 2: These three actually all have separate precedence in CPP

```suggestion
|      |    ¦    |   e1 | e2    |   left   |    Fortran IOR(e1, e2)     |
|------+---------+--------------+----------+----------------------------|
|      |    ^    |   e1 ^ e2    |   left   |    Fortran IAND(e1, e2)    |
|------+---------+--------------+----------+----------------------------|
|      |    &    |   e1 & e2    |   left   |    Fortran IEOR(e1, e2)    |
```

|------+---------+--------------+----------+----------------------------|

> |      |   ==    |   e1 == e2   |   left   |  1 if e1 == e2, 0 otherwise |

Reviewer: ```suggestion
|      |   ==    |   e1 == e2   |   left   | 1 if e1 == e2, 0 otherwise |
```

|      |   !=    |   e1 != e2   |   left   | 1 if e1 /= e2, 0 otherwise |
|------+---------+--------------+----------+----------------------------|

> |      |    >    |   e1 > e2    | nonassoc | 1 if e1 > e2, 0 otherwise  |
> |      |   >=    |   e1 >= e2   | nonassoc | 1 if e1 >= e2, 0 otherwise |
> |      |    <    |   e1 < e2    | nonassoc | 1 if e1 < e2, 0 otherwise  |
> |      |   <=    |   e1 <= e2   | nonassoc | 1 if e1 <= e2, 0 otherwise |

Reviewer 2: In C (and hence CPP) _relational-expression_ operators are left associative.

Example:
```
#if 0 > -4 > 0
left-assoc
#else
right-assoc
#endif
```

Expands to "left-assoc" in CPP.

However it appears some FPPs may (unintentionally?) [diverge on this detail](https://godbolt.org/z/Yfzc4hnr5). Another great example why we need standardization!

```suggestion
|      |    >    |   e1 > e2    |   left   | 1 if e1 > e2, 0 otherwise  |
|      |   >=    |   e1 >= e2   |   left   | 1 if e1 >= e2, 0 otherwise |
|      |    <    |   e1 < e2    |   left  | 1 if e1 < e2, 0 otherwise  |
|      |   <=    |   e1 <= e2   |   left  | 1 if e1 <= e2, 0 otherwise |
```

Author: I put them as non-associative in the Fortran preprocessor, for potential future compatibility if we ever add Fortran operators. The Fortran operators are non-associative, so we might want to disallow expressions like `0 > -4 > 0` now. (And I think chaining relational operators is terrible, but that's just my opinion. If I saw that in C code myself, I would probably replace it with something people could immediately understand.)

|------+---------+--------------+----------+----------------------------|
|      |   <<    |   e1 << e2   |   left   | Fortran ISHFT(e1, e2)      |
|      |   >>    |   e1 >> e2   |   left   | Fortran ISHFT(e1, -e2)     |
|------+---------+--------------+----------+----------------------------|
|      |    +    |   e1 + e2    |   left   |             +              |
|      |    -    |   e1 - e2    |   left   |             -              |
|------+---------+--------------+----------+----------------------------|
|      |    *    |   e1 * e2    |   left   |             *              |
|      |    /    |   e1 / e2    |   left   |             /              |
|      |    %    |   e1 % e2    |   left   |    Fortran MOD(e1, e2)     |
|------+---------+--------------+----------+----------------------------|

> |      | unary + |     + e1     |  right   |        unary +             |
> |      | unary + |     - e2     |  right   |        unary -             |

Reviewer 2: You forgot bitwise NOT and logical NOT

```suggestion
|      | unary + |     + e1     |  right   |        unary +             |
|      | unary - |     - e1     |  right   |        unary -             |
|      | unary ~ |     ~ e1     |  right   |     Fortran NOT(e1)        |
|      | unary ! |     ! e1     |  right   | 1 if e1 == 0, 0 otherwise  |
```

|      | defined |  defined ID  | nonassoc |    1 if the identifier     |
|      |         |              |          |   has a #defined value,    |

> |      |         |              |          |     0 otherwise            |

Reviewer 2: On further thought, this is just plain wrong. 
`defined` cannot appear in this table, because it needs to be applied AFTER macro expansion and BEFORE ID replacement with zero. Hence the `defined` operator (as in CPP) must be resolved and replaced before this post-expansion grammar is applied.

at different precedence levels.

An "associativity" is assigned to each operator that determines how
operators at the same precedence level are combined.
    - "left" means the operator binds to the left,
    - "right" means the operator binds to the right.
    - "nonassoc" means that the operator is not associative.

The following table describes the semantics of the allowed operators
in conditional expressions. The table is grouped by precedence level,
from lowest precedence to highest.

We label subexpressions "e1", "e2", and "e3" to aid in describing the
evaluation semantics. Unless otherwise specified, all operators
evaluate with the same semantics as their Fortran counterparts.

| Prec |   Op    |    Syntax    |  Assoc'y |  Evaluation Semantics      |
|------+---------+--------------+----------+----------------------------|
| low  |   ? :   | e1 ? e2 : e3 |  right   |      conditional-expr      |
|------+---------+--------------+----------+----------------------------|
|      |   ¦¦    |   e1 || e2   |   left   |       Fortran .OR.         |
|------+---------+--------------+----------+----------------------------|

> |      |   &&    |   e1 && e2   |   left   |       Fortran .AND.        |

Reviewer 2: In the 5/12 call we resolved these should be short-circuit evaluation as in CPP, to allow things like:
```
#if  x && 1/x
#endif
```
which means they are NOT simply  Fortran .OR. / .AND.
